### ISSUE TITLE
Change the timeout example to use context

### ISSUE THREAD

**Issue by mdaffin** [mmcgrana/gobyexample#207]
There are some major flaws with the timeouts examples notably;

* It does not stop the processing of the function
* It leaks a channel that blocks forever
* It leaks a goroutine that sleeps forever waiting on the above channel

This makes it a fairly bad example to use in any form of production workload and should really be avoided in all workloads as it teaches some bad practices. It is far better to use the context package for timeouts. 

Would you be open to a pull request changing the example to use the context package instead? And using a now far more idiomatic pattern that does not suffer from the above problems?

I see far too many people linking to this example completely unaware of its flaws and thus using it in situations that just cause a large amount of problems.

**Comment by eliben** [comment]
Can you please clarify where you see the leaks happening in the specific example here? Note that "channel leaks" is not really a thing - it's perfectly fine to leave channels open at the end of programs (they are not finite file descriptors).

FWIW we should also add an example of using context cancellation, but this example demonstrates an interesting use of timers with `select` in a very simple way, which I think is still useful. It is hopefully implied that basic samples are not always suitable for "production" uses; on the other hand, production-proofing every sample will make trivial examples very hard to do.

**Comment by mdaffin** [comment]
>  it's perfectly fine to leave channels open at the end of programs 

The problem is when you use the example in the middle of a long running application. A channel and go routine costs a small amount of memory which over time builds up to a none trivial amount if you do not allow them to be cleaned up after you are done with them. If every program exited straight away it would not be an issue but go, and this pattern, is often used in the middle of long running servers which introduces problems over time.

Or worst, when people use this pattern to try and cancel computation, which you cannot do. With this pattern you can leak far more expensive resources then a simple channel or go routine.

We now have [context](https://golang.org/pkg/context/) which is a far better for asking gorouteines to stop then the patterns used in this example.

> It is hopefully implied that basic samples are not always suitable for "production" uses

Examples might not be production quality, which is ok, but neither is the pattern they are trying to teach - which is the problem. Examples should be easy to extend to production code, the patterns introduced here should not be used in production which makes them of no real values and, if anything, an anti-pattern.

It just ends up being a trap for newer users to fall into when they are learning go. Worst, a trap they only learn about when trying to use these patterns in production code. We have better patterns, we should be teaching them.

**Comment by eliben** [comment]
@mdaffin you still haven't answered "Can you please clarify where you see the leaks happening in the specific example here?"

> Or worst, when people use this pattern to try and cancel computation, which you cannot do.

There's nothing magical about contexts that would let you "cancel a computation" more so than the `select` shown here. Canonical examples of contexts with timeouts have the same `select`.

But I do agree we should showcase contexts too - this is planned, actually. Feel free to propose a "timeout using channel" example in a PR. I think it would be a useful addition, and we can definitely point to it from this example.

**Comment by eliben** [comment]
To propose a path forward, I suggest:

1. Enhance the comment in `timeouts.go` mentioning the buffered channel that ensures no goroutine leak - this is a common Go practice for these cases
2. Open a separate issue for adding an example of context. Then this timeouts example could link there for "something more advanced"

Then we should be able to close this issue

**Comment by mmcgrana** [comment]
@eliben that sounds good to me.

**Comment by eliben** [comment]
For (1) sent #271 

For (2) opened #273 

Will close this issue once #271 lands
